### PR TITLE
Handle associated type shorthand (`T::Item`) (Second attempt)

### PR DESCRIPTION
**Pull request by flodiebold**
This is only allowed for generic parameters (including `Self` in traits), and
special care needs to be taken to not run into cycles while resolving it,
because we use the where clauses of the generic parameter to find candidates for
the trait containing the associated type, but the where clauses may themselves
contain instances of short-hand associated types.

In some cases this is even fine, e.g. we might have `T: Trait<U::Item>, U:
Iterator`. If there is a cycle, we'll currently panic, which isn't great, but
better than overflowing the stack...

### CODE REVIEW

**Comment by matklad** [comment]
@flodiebold note that salsa supports recovery from cycle error now: https://github.com/salsa-rs/salsa/blob/a9860bf37f08c2b8029d62aacca60ea43c35d767/tests/cycles.rs

File: crates/ra_hir/src/ty/lower.rs [hunk 1]
```diff
@@ -491,6 +547,29 @@ pub(crate) fn type_for_field(db: &impl HirDatabase, field: StructField) -> Ty {
...
+    def: GenericDef,
+    param_idx: u32,
+) -> Arc<[GenericPredicate]> {
+    let resolver = def.resolver(db);
+    let predicates = resolver
+        .where_predicates_in_scope()
+        // we have to filter out all other predicates *first*, before attempting to lower them
+        .filter(|pred| Ty::from_hir_only_param(db, &resolver, &pred.type_ref) == Some(param_idx))
+        .flat_map(|pred| GenericPredicate::from_where_predicate(db, &resolver, pred))
+        .collect::<Vec<_>>();
```
Comment (matklad) [inline crates/ra_hir/src/ty/lower.rs]: Ok, this time I am sure that we can just collect into an Arc slice: https://github.com/rust-lang/rust/blob/master/RELEASES.md#libraries

**Comment by matklad** [comment]
LGTM, but maybe it makes sense to kick the tiers of the new salsa cycle recovery, while we are at it?

File: crates/ra_hir/src/ty/lower.rs (hunk 1 above)
Comment (flodiebold) [reply crates/ra_hir/src/ty/lower.rs]: That releases on the 26., so we still have to wait a few days ;)

**Comment by flodiebold** [comment]
The cycle recovery isn't released yet, is it?

**Comment by flodiebold** [comment]
bors r+

I'll do the cycle recovery once Salsa is updated :)

**Pull request merged**